File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/1654-랜선 자르기/solution.py
import sys
si = sys.stdin.readline
k, n = list(map(int, si().split()))
a = []
for i in range(k):
    a.append(int(si()))

def determination(len):
    sum = 0
    for x in a:
        sum += x // len
    return sum >= n

l, r, ans = 1, 1 << 31, 0
while l <= r:
    mid = (l + r) // 2
    if determination(mid):
        ans = mid
        l = mid + 1
    else:
        r = mid - 1

print(ans)

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/1654-랜선 자르기/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 10005

int N, K;
int A[NM];

void input() {
	cin >> K >> N;
	for (int i = 1; i <= K; i++) cin >> A[i];
}

bool determination(ll len) {
	ll sum = 0;
	for (int i = 1; i <= K; i++) {
		sum += A[i] / len;
	}
	return sum >= N;
}

void pro() {
	ll L = 1, R = 1ll << 31, ans = 0, ansCnt = 0;
	while (L <= R) {
		ll mid = (L + R) / 2;
		if (determination(mid)) {
			ans = mid;
			L = mid + 1;
		}
		else R = mid - 1;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}